scylla-cql/src/types/serialize/row.rs

```diff
@@ -125,13 +131,150 @@ pub enum ValueListToSerializeRowAdapterError {
     NoBindMarkerWithName { name: String },
 }
 
+/// A container for already serialized values. It is not aware of the types of contained values,
+/// it is basically a byte buffer in the format expected by the CQL protocol.
+/// Usually there is no need for a user of a driver to use this struct, it is mostly internal.
+/// The exception are APIs like `ClusterData::compute_token` / `ClusterData::get_endpoints`.
+/// Allows adding new values to the buffer and iterating over the content.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NewSerializedValues {
+    serialized_values: Vec<u8>,
+    element_count: u16,
+}
+
+impl NewSerializedValues {
+    pub const fn new() -> Self {
+        NewSerializedValues {
+            serialized_values: Vec::new(),
+            element_count: 0,
+        }
+    }
+
+    pub fn from_serializable<T: SerializeRow>(
+        ctx: &RowSerializationContext,
+        row: &T,
+    ) -> Result<Self, SerializationError> {
+        let mut data = Vec::new();
+        let element_count = {
+            let mut writer = BufBackedRowWriter::new(&mut data);
+            T::preliminary_type_check(ctx)?;
+            row.serialize(ctx, &mut writer)?;
+            writer.value_count()
+        };
+
+        Ok(NewSerializedValues {
+            serialized_values: data,
+            element_count,
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.element_count() == 0
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = RawValue> {
+        NewSerializedValuesIterator {
+            serialized_values: &self.serialized_values,
+        }
+    }
+
+    pub fn element_count(&self) -> u16 {
+        // We initialize first two bytes in new() and BufBackedRowWriter does too,
+        // so this unwrap is safe
+        self.element_count
+    }
+
+    pub fn buffer_size(&self) -> usize {
+        self.serialized_values.len()
+    }
```

Comment:
Those can be inline, too. Remember that non-generic functions are not automatically inlined across crates without the annotation. The functions here are quite trivial, so they would benefit from inlining. The drawback is longer compilation time, but I don't think it should matter as much for such simple functions.